### PR TITLE
feat: literal expressions for the textplan parser

### PR DESCRIPTION
**Pull request by EpsilonPrime**
* Adds support for literals and column names on the parser side
      * Adds the datetime submodule for parsing dates 
   * Initial output to binary plans
      * Adds the protobuf-matchers submodule for testing

### CODE REVIEW

**Review by westonpace — COMMENTED** [review]
Need to keep looking but here are some notes from our brief in person review

File: src/substrait/textplan/parser/tests/TextPlanParserTest.cpp [hunk 1]
```diff
@@ -207,64 +220,453 @@ std::vector<TestCase> getTestCases() {
...
+              R"(relations { rel { project {
+              expressions { literal { null: { fp32 { nullability: NULLABILITY_NULLABLE } } } }
+              expressions { literal { null: { fp64 { nullability: NULLABILITY_NULLABLE } } } }
+              } } })"))),
+      },
+      {
+          "test10-literals-decimal",
+          R"(project relation literalexamples {
+            expression 42_decimal<5,4>;
+            expression 42_decimal<5,-4>;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/tests/TextPlanParserTest.cpp:310]: It was hidden as an ASCII character.  I've replaced the ASCII with octal so it is more clear.

File: src/substrait/textplan/parser/tests/TextPlanParserTest.cpp [hunk 2]
```diff
@@ -207,64 +220,453 @@ std::vector<TestCase> getTestCases() {
...
+          "test10-literals-lists",
+          R"(project relation literalexamples {
             expression {"a", "b", "c"}_list<string>;
             expression {null, "a", "b"}_list<string?>;
-            expression {42 : "life", 32 : "everything"}_map<int, string>;
             expression {}_list<string>;
             expression {}_list<string?>;
-            expression {}_list?<string?>;
-            expression {}_map<int, string>;
+            expression {}_list?<string>;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/tests/TextPlanParserTest.cpp:491]: Done.

File: src/substrait/expression/DecimalLiteral.cpp [hunk 3]
```diff
@@ -27,13 +28,38 @@ std::string negate(const std::string& value) {
...
+    int32_t scale) {
+  absl::uint128 v;
+  if (!absl::SimpleAtoi(str, &v)) {
+    // TODO -- Store the parse errors so that they can be examined later.
+    return {"", 0, 0};
+  }
+  std::uint8_t valueBytes[16];
+  uint128ToBytes(v, valueBytes);
+  // TODO -- Store parse errors locally so folks can determine what went wrong.
+  return {std::string((const char*)valueBytes, 16), precision, scale};
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/expression/DecimalLiteral.cpp:60]: Removed.

File: src/substrait/textplan/parser/ParseText.cpp [hunk 4]
```diff
@@ -75,11 +75,25 @@ ParseResult parseStream(antlr4::ANTLRInputStream stream) {
...
+  auto relationVisitor = std::make_shared<SubstraitPlanRelationVisitor>(
+      *pipelineVisitor->getSymbolTable(), pipelineVisitor->getErrorListener());
+  try {
+    relationVisitor->visitPlan(tree);
+  } catch (...) {
+    parserErrorListener.syntaxError(
+        &parser,
+        nullptr,
+        1,
+        1,
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/ParseText.cpp:87]: Updated.

File: src/substrait/textplan/parser/SubstraitPlanRelationVisitor.cpp [hunk 5]
```diff
@@ -0,0 +1,1238 @@
...
+      type->mutable_map()->set_nullability(
+          ::substrait::proto::Type_Nullability_NULLABILITY_NULLABLE);
+      break;
+    case ::substrait::proto::Type::kUserDefined:
+      type->mutable_user_defined()->set_nullability(
+          ::substrait::proto::Type_Nullability_NULLABILITY_NULLABLE);
+      break;
+    case ::substrait::proto::Type::kUserDefinedTypeReference:
+    case ::substrait::proto::Type::KIND_NOT_SET:
+      break;
```
[the remark this answers is not in the document]
Comment (EpsilonPrime) [reply src/substrait/textplan/parser/SubstraitPlanRelationVisitor.cpp:144]: I tested out adding an error here and it ended up being duplicative with the "Unable to recognize requested type." error message which identifies the issue earlier (but continues processing to find other errors).  So I added a comment that an error is not needed.

**Review by westonpace — APPROVED** [review]

**Pull request merged**